File: 0045-jump-game-ii/0045-jump-game-ii.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        int minJumps = 0;
        int currentEnd = 0;
        int farthest = 0;

        // We don't need to check the last element because we are already
        // jumping to it.
        for (int i = 0; i < nums.size() - 1; i++) {
            farthest = max(farthest, i + nums[i]);

            // We are allowed to jump and we have reached the end of our current
            // jump, so we have to make another jump.
            if (i == currentEnd) {
                minJumps++;
                currentEnd = farthest;

                // If our current jump can reach the end, we break the loop.
                if (currentEnd >= nums.size() - 1) {
                    break;
                }
            }
        }

        return minJumps;
    }
};

File: 0045-jump-game-ii/0045-jump-game-ii.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        minjumps = 0
        farthest = 0
        currend = 0
        for i in range(len(nums)-1):
            farthest = max(farthest,nums[i]+i)

            if i == currend:
                currend = farthest
                minjumps += 1
            
            if currend >= len(nums) - 1:
                break
        return minjumps